news/_posts/2019-21-01-resilient-php-applications.md

```diff
@@ -0,0 +1,105 @@
+---
+layout: post
+title:  "Resilient PHP applications"
+subtitle: "Introducing the PrestaShop Circuit Breaker library"
+date:   2019-01-21 10:00:00
+authors: [ mandrieu ]
+icon: icon-blaster
+tags:
+ - 1.7
+---
+
+No matter how well designed is your project and how many tests you have, at some point your website can become unavailable.
+
+This can be because of your database, or calls to third party services are unavailable (e.g: payment methods API, PrestaShop REST API to retrieve modules information, synchronization with an ERP, automatisation of warehouse & logistic, ...)
+
+Instead of thinking everything will be ok, we need - as developers - to manage the failures of our systems.
+
+We need to create resilient systems, to create websites able to recover quickly from network disasters.
+
+## The risks of depending on external services
+
+Assuming all external services will works fine is risky because you don't think about what should happens in case of unavailability:
+
+* if the payment APIs are not reachable, instead of freezing your shop you could suggest your customer to contact them back once the system is ok and provide them a little discount;
+
+* you should be able to install PrestaShop even if the translations APIs are not reachable,
+at least in english;
+
+* you should be able to access your back office, no matter how many external APIs are not available;
+
+**The collapse of a third party service shouldn't collapse your shop.**
+
+Another issue with this kind of naive assumptions is that your system will try to contact again and again the third party service. This is slowing your system and make the user experience terrible when you could manage to provide an alternative response.
+
+> There is no good reason to provide such a terrible user experience! period.
+
+## Introducing the PrestaShop Circuit Breaker library
+
+The [Circuit Breaker](https://martinfowler.com/bliki/CircuitBreaker.html) design pattern provides strategies and functions to manage the unavailability of the third party services.
+
+The basic idea behind the circuit breaker is very simple:
+
+* every call of third party service is wrapped into a function call from the Circuit Breaker;
+* once we reach a defined numbers of failures (means the service is unreachable), the Circuit Breaker is "Open": an alternative response is provided and no calls to the service are done;
+* after a defined amount of time (called "reset timeout" sometimes), the Circuit Breaker try to access the service: if the service is available it will be used again, else the alternative response is provided instead;
+
+![https://martinfowler.com/bliki/images/circuitBreaker/sketch.png](https://martinfowler.com/bliki/images/circuitBreaker/sketch.png)
```

Comment:
@mickaelandrieu need the link to the original blog article, not only the picture. could you please provide it?